widgets/matplotlib/simulation/recoil_atom_distribution.py

```diff
@@ -822,14 +826,75 @@ def open_recoil_element_info(self):
                         self.current_recoil_element)
                 self.update_recoil_element_info_labels()
                 self.update_colors()
-            except KeyError:
-                error_box = QtWidgets.QMessageBox()
-                error_box.setIcon(QtWidgets.QMessageBox.Warning)
-                error_box.addButton(QtWidgets.QMessageBox.Ok)
-                error_box.setText(
-                    "All recoil element information could not be saved.")
-                error_box.setWindowTitle("Error")
-                error_box.exec()
+
+            except KeyError as e:
+                print(str(e))
+                QtWidgets.QMessageBox.critical(
+                    self,
+                    "Error",
+                    "All recoil element "
+                    "information could not be saved.",
+                    QtWidgets.QMessageBox.Ok,
+                    QtWidgets.QMessageBox.Ok)
+                return
+
+    def open_recoil_element_info(self):
+        """Open recoil element info.
+        """
+
+        dialog = ReferenceDensityDialog(
+            self.current_recoil_element, self.current_element_simulation)
+
+        if self.current_recoil_element is None:  # Otherwise crash
+            return
+
+        if dialog.userSelectionCheckBox.isChecked():
+
+            self.current_recoil_element.manual_reference_density_checked = True
+
+            self.parent.targetUserSelectionLabel.setText(
+                "Using user-defined reference density")
+            self.parent.targetUserSelectionLabel.setStyleSheet(
+                "background-color: yellow")
+
+            self.parent.recoilUserSelectionLabel.setText(
+                "Using user-defined reference density")
+            self.parent.recoilUserSelectionLabel.setStyleSheet(
+                "background-color: yellow")
+        else:
+            self.current_recoil_element.manual_reference_density_checked = False
+            self.parent.targetUserSelectionLabel.setText("")
+            self.parent.recoilUserSelectionLabel.setText("")
+
+        if dialog.isOk:
+            try:
+                new_values = dialog.get_properties()
+                old_recoil_name = self.current_recoil_element.name
+                self.current_element_simulation.update_recoil_element(
+                    self.current_recoil_element,
+                    new_values)
```

Comment:
`dialog.get_properties()` does not return everything that's needed for `RecoilElement.update()` (name, description, new_values) anymore. This causes a `KeyError` in `RecoilElement.update()`.

A possible fix would be to only update the reference density.

Also, same thing as in my upper comment: this reference density should be changed for all `RecoilElement`s in the simulation. This should probably be done in #236 because it'll move the reference density from individual `RecoilElement`s to a more central location.

Reply:
Yes